ccm_web/server/src/lti/lti.service.ts

```diff
@@ -45,7 +47,31 @@ export class LTIService implements BeforeApplicationShutdown {
 
     // Redirect to the application root after a successful launch
     provider.onConnect(async (token: IdToken, req: Request, res: Response) => {
-      logger.debug(JSON.stringify(token.userInfo))
+      logger.debug(`LTI launch is successfull! ${JSON.stringify(token.userInfo)}`)
+      const customLTIVariables = token.platformContext.custom
+      if (customLTIVariables == null) {
+        return res.json({ lti_error: 'LTI launch is missing custom attributes, please add it in LTI configutation in Canvas' })
+      }
+      const loginId = customLTIVariables.login_id as string
+      const [record, created] = await this.userService.upsertUser({
+        firstName: token.userInfo.given_name,
+        lastName: token.userInfo.family_name,
+        email: token.userInfo.email,
+        loginId: loginId,
+        ltiId: token.user
+      })
+      /* exclusively treating null value as false, But in really the const `created` won't be null with mysql dialect.
+      This is typescript being smart. See more comments on this in user.service.ts */
+      logger.info(
+        `User ${record.loginId} is ${
+          (created ?? false) ? 'created' : 'updated'
+        } in 'user' table`
+      )
+      // if the user inserting into DB failed don't want to proceed further
+      if (!(record instanceof User)) {
+        logger.error(`something went wrong in creating user: ${loginId}`)
+        return res.status(HttpStatus.INTERNAL_SERVER_ERROR).json('The Launch of application failed, please try to refresh the page.')
+      }
```

Comment:
Were you able to trigger this code branch? Not sure how you would do that, so wondering  if it's necessary, or if other error/Promise handling would take care of this.

Reply:
So, I experimented a little and couldn't find/think of a way to do this. The method definition says it will return a model instance, so I think the only way it wouldn't is if an exception was thrown while trying to connect to/modify the database, which would bubble up separately through promise handling. I was able to do that by messing with the database parameters in `app.module.ts`. Maybe there are other ways an exception would be thrown from an invalid record (which are hard to reproduce because we're typing everything), but again, we wouldn't have a `record` to check here. I think if you want to do this, we need to use a `try`/`catch` statement, doing something like this:
```
      try {
        const [record, created] = await this.userService.upsertUser({
          firstName: token.userInfo.given_name,
          lastName: token.userInfo.family_name,
          email: token.userInfo.email,
          loginId: loginId,
          ltiId: token.user
        })
        logger.info(
          `User ${record.loginId} is ${
            (created ?? false) ? 'created' : 'updated'
          } in 'user' table`
        )
        return provider.redirect(res, '/')
      } catch (error) {
        logger.error(`something went wrong in creating user: ${loginId}`)
        logger.error(`exception: ${String(error)}`)
        return res.json('The application failed to launch for this user; please contact support.')
      }
```

This raises some questions about error handling in general for database calls...

Reply:
To me what makes sense is to surround the DB [call](https://javascript.info/async-await#error-handling) around the try catch. I would lie if I say i did not thought about this. But When I looked the documentation of Sequelize for upserting method, it only said it will

Reply:
I resolved you concern with handling of DB failures  [here](https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/80/commits/4e021857d5a6cafde5ec52f3145444441aa4e9d8)